src/Conversion/ONNXToKrnl/Quantization/DynamicQuantizeLinear.cpp

```diff
@@ -0,0 +1,155 @@
+  ONNXDynamicQuantizeLinearOpLowering(
+      TypeConverter &typeConverter, MLIRContext *ctx)
+      : OpConversionPattern(typeConverter, ctx) {}
+
+  LogicalResult matchAndRewrite(ONNXDynamicQuantizeLinearOp dqlOp,
+      ONNXDynamicQuantizeLinearOpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    using LocalDialectBuilder = MultiDialectBuilder<KrnlBuilder,
+        IndexExprBuilderForKrnl, MathBuilder, MemRefBuilder, OnnxBuilder>;
+    Operation *op = dqlOp.getOperation();
+    Location loc = ONNXLoc<ONNXDynamicQuantizeLinearOp>(op);
+    LocalDialectBuilder create(rewriter, loc);
+
+    ValueRange operands = adaptor.getOperands();
+    Value X = adaptor.getX();
+
+    // MemRefType for inputs and outputs.
+    auto xMemRefType = dyn_cast<MemRefType>(X.getType());
+    auto yMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(dqlOp.getResult(0).getType()));
+    auto yScaleMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(dqlOp.getResult(1).getType()));
+    auto yZeroPointMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(dqlOp.getResult(2).getType()));
+
+    // Types
+    Type elementType = xMemRefType.getElementType();
+    Type quantizedElementType = yMemRefType.getElementType();
+    int64_t rank = xMemRefType.getRank();
+
+    // Get shape.
+    ONNXDynamicQuantizeLinearOpShapeHelper shapeHelper(
+        op, operands, &create.krnlIE);
+    shapeHelper.computeShapeAndAssertOnFailure();
+
+    // Allocate output buffers.
+    Value Y =
+        create.mem.alignedAlloc(yMemRefType, shapeHelper.getOutputDims(0));
+    Value YScale =
+        create.mem.alignedAlloc(yScaleMemRefType, shapeHelper.getOutputDims(1));
+    Value YZeroPoint = create.mem.alignedAlloc(
+        yZeroPointMemRefType, shapeHelper.getOutputDims(2));
+
+    // Equations:
+    // y_scale = (max(x) - min(x))/(qmax - qmin)
+    // intermediate_zero_point = qmin - min(x)/y_scale
+    // y_zero_point = cast(round(saturate(itermediate_zero_point)))
+    // y = saturate (round (x / y_scale) + y_zero_point)
+    //
+    // where, saturate is to clip to [0, 255] for ui8.
+
+    // QMax, QMin.
+    Value qMax = create.math.constant(elementType, 255.0);
+    Value qMin = create.math.constant(elementType, 0.0);
+    Value QMax = create.mem.alignedAlloc(yScaleMemRefType);
+    create.krnl.store(qMax, QMax);
+    Value QMin = create.mem.alignedAlloc(yScaleMemRefType);
+    create.krnl.store(qMin, QMin);
+
+    // Compute max(x) and min (x).
+    Value none = create.onnx.none();
+    Value XMax = create.onnx.toMemref(
+        create.onnx.reduceMax(yScaleMemRefType, X, none, false));
+    Value XMin = create.onnx.toMemref(
+        create.onnx.reduceMin(yScaleMemRefType, X, none, false));
+    Value xMax = create.krnl.load(XMax);
+    Value xMin = create.krnl.load(XMin);
+    // Include 0 to max(x) and min(x).
+    // x_min = min(min(x), 0)
```

Comment:
This is to include 0 as one element in the X tensor when computing `min(X)`. Same to `max`.
I found that this is necessary to get correct result. DynamicQuantizeLinear spec does not mention this, but its python example does: https://github.com/onnx/onnx/blob/main/docs/Operators.md#DynamicQuantizeLinear
```python
X = np.array([0, 2, -3, -2.5, 1.34, 0.5]).astype(np.float32)
x_min = np.minimum(0, np.min(X))
x_max = np.maximum(0, np.max(X))
```

Reply:
I bumped into this today. As @tungld mentioned, the [DynamicQuantizeLinear Doc](https://onnx.ai/onnx/operators/onnx__DynamicQuantizeLinear.html) page doesn't show this (but I think it's alluded to by `data range is adjusted to include 0.`). The reference example does have it at https://github.com/onnx/onnx/blob/main/onnx/reference/ops/op_dynamic_quantize_linear.py#L15-L16.

I did a little digging, and it looks like PyTorch does the same thing: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/QTensor.cpp quantize_per_tensor_dynamic() -> quant_utils::ChooseQuantizationParams -> https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/QuantUtils.h#L95-L99
```
  // We extend the [min, max] interval to ensure that it contains 0.
  // Otherwise, we would not meet the requirement that 0 be an exactly
  // representable value.
  min = std::min(min, 0.f);
  max = std::max(max, 0.f);
```

I created https://github.com/onnx/onnx/issues/5597 to hopefully get the doc updated to match the reference example.